### PR TITLE
[wasm][interp] Back to recursive, but split so recursive function is small.

### PR DESCRIPTION
**Pull request by jaykrell**
This has been brought up and implemented countless times already.

The latest reason for this, is that the non-recursive interpreter
stresses JITs, in that it requires either:
 - a first tier optimizing JIT -- which is expected on iOS soon (now?)
 - or on-stack-replacement -- which is expected to exists, but maybe not, or benchmarks miss it?
    And possibly current-function replacement? Does that occur?
 - or maybe a warm-up phase -- seems bad

This improves WebAssembly performance by 12% as measured with
`run-benchmark-wasm.sh bilinear`.

This has been verified to also address the stack overflow problem
 with current Blazor release from around early January 2020,
 which master's non-recursive interpreter also does.

I will record here some of the benchmarks.
These appear to avoid the tiering issue, but yet show this to be faster.
This data all hand transcribed instead of copy/pasted, sorry.

master (https://github.com/mono/mono/commit/488da04761f74c3bde5ac52d04454425b27ffdd6):
 bilinear avg:27097 min:27053 max:27200
 segmentation avg:17668 min:17623 max:17720
 poker avg:6981 min:6953 max:7035
 exceptions avg:0.0514 min:0.0457 max:0.3736

PR (from https://github.com/mono/mono/commit/9ce7798c41f531c366a348492bf474e1b29168d9):
 bilinear avg:23568 min:23509 max:23749 (avg 13% faster)
 segmentation avg:14535 min:14406 max:14898 (avg 17% faster)
 poker avg:6079 min:6038 max:6144 (avg 13% faster)
 exceptions avg:0.0356 min:0.0334 max:0.1859 (avg 30% faster)

For non-wasm, non-recursive is faster.

master poker avg:1345 1546 1501 (multiple runs)
PR poker avg:1662 1737 1699 1790 (multiple runs)

master bilinear avg:4317
PR bilinear avg:6178 6104

master segmentation avg:3117 3107 (multiple runs)
PR segmentation avg:3910 3690 (multiple runs)

master exceptions avg:0.00459
PR exceptions avg:0.00446

### CODE REVIEW

**Comment by jaykrell** [comment]
/cc @kg 
Maybe try the benchmarks with this?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
/cc @lewing @vargaz @kg @lewurm @BrzVlad
12-30% benchmark improvement worthwhile?
Do more benchmarking? Don't like the code no matter how fast it is?

File: mono/mini/interp/interp-internals.h [hunk 1]
```diff
@@ -176,31 +176,20 @@ struct InterpMethod {
...
-
 /* Arguments that are passed when invoking only a finally/filter clause from the frame */
 typedef struct FrameClauseArgs FrameClauseArgs;
 
+#define INTERP_STATE		\
+	const guint16 *ip;	\
+	stackval *sp;		\
+	guchar *vt_sp;		\
+	GSList *finally_ips;	\
+	guchar *vtalloc; /* only for DEBUG_INTERP */
```
Comment (jaykrell) [inline mono/mini/interp/interp-internals.h]: A change is lost here, alas.

**Comment by lambdageek** [comment]
Going to hold off on this on advice from the WASM team

**Comment by jaykrell** [comment]
Now reported that Firefox/SpiderMonkey does not on-stack-replace, further motivating this PR.

**Pull request closed without merging**